### PR TITLE
FIX: [grid2] use tryLock during the openGrid process

### PR DESCRIPTION
**Pull request by c9s**

### CODE REVIEW

File: pkg/types/ordermap.go [hunk 1]
```diff
@@ -125,28 +125,36 @@ func (m *SyncOrderMap) Remove(orderID uint64) (exists bool) {
...
+	if len(m.pendingRemoval) == 0 {
+		return
+	}
+
 	m.Lock()
 	defer m.Unlock()
 
-	m.orders.Add(o)
-
-	if len(m.pendingRemoval) > 0 {
```
Comment (c9s) [inline pkg/types/ordermap.go:134]: move code block to processPendingRemoval

File: pkg/bbgo/order_execution.go [hunk 2]
```diff
@@ -340,6 +341,8 @@ func BatchRetryPlaceOrder(ctx context.Context, exchange types.Exchange, errIdx [
 		createdOrders, errIdx, err2 = BatchPlaceOrder(ctx, exchange, orderCallback, submitOrders...)
 		if err2 != nil {
 			werr = multierr.Append(werr, err2)
+		} else if err2 == nil {
```
Comment (gx578007) [inline pkg/bbgo/order_execution.go:344]: just using `else {`?

Comment (c9s) [reply pkg/bbgo/order_execution.go:344]: oh yes

File: pkg/strategy/grid2/strategy.go [hunk 3]
```diff
@@ -1179,7 +1179,11 @@ func (s *Strategy) updateFilledOrderMetrics(order types.Order) {
 }
 
 func (s *Strategy) updateGridNumOfOrdersMetricsWithLock() {
-	s.updateGridNumOfOrdersMetrics(s.getGrid())
+	if s.mu.TryLock() {
```
[the remark this answers is not in the document]
Comment (gx578007) [reply pkg/strategy/grid2/strategy.go:1182]: Could it log some messages when failing to acquire the lock?

[the remark this answers is not in the document]
Comment (c9s) [reply pkg/strategy/grid2/strategy.go:1182]: Sure

File: pkg/types/ordermap.go [hunk 4]
```diff
@@ -125,28 +125,36 @@ func (m *SyncOrderMap) Remove(orderID uint64) (exists bool) {
 	return exists
 }
 
-func (m *SyncOrderMap) Add(o Order) {
+func (m *SyncOrderMap) processPendingRemoval() {
+	if len(m.pendingRemoval) == 0 {
```
Comment (gx578007) [inline pkg/types/ordermap.go:129]: `pendingRemoval` should be protected by the lock as well?

Comment (c9s) [reply pkg/types/ordermap.go:129]: it’s read only, should be okay?

**Review by gx578007 — APPROVED** [review]

**Pull request merged**